### PR TITLE
Handle duplicates in destructuring (Fix #200)

### LINKED ISSUE #200: destruct-param conflicts with variables introduced by itself
When transforming a function with two objects that have the same field:

```js
function fn(a, b) {
  console.log(a.foo, b.foo);
}
```

Lebab produces the following:

```js
function fn({foo}, {foo}) {
  console.log(foo, foo);
}
```

The same problem happens when there's an inner function that has object parameter with the same field:

```js
function fn(a) {
  function inner(b) {
    console.log(a.foo, b.foo);
  }
}
```

Again Lebab produces:

```js
function fn({foo}) {
  function inner({foo}) {
    console.log(foo, foo);
  }
}
```

`destructParamTest.js` contains skipped tests for these scenarios.

### PR DESCRIPTION
**Pull request by SindreSvendby**
This fixes https://github.com/lebab/lebab/issues/200

I really liked this project, could you maybe label some of the issue as [good first issues](https://github.com/babel/babel/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)   like babel does? Or do you have any other suggestion on what you want me to do if I want to contribute some to this?

### CODE REVIEW

File: src/transform/destructParam.js [hunk 1]
```diff
@@ -25,6 +25,17 @@ export default function(ast, logger) {
               return;
             }
 
+            // skip destructuring, if one of the destructured variable names is already declared by another destruction
+            const duplicateDetected = exs.some(ex => {
+              if (variableNamesCreated.includes(ex.property.name)) {
+                return true;
+              }
+              return false;
+            });
```
Comment (uniibu) [inline src/transform/destructParam.js]: You can simplify this to
```js
const duplicateDetected = exs.some(ex => {
  return variableNamesCreated.includes(ex.property.name);
});
```
or
```js
const duplicateDetected = exs.some((ex) => variableNamesCreated.includes(ex.property.name));
```
But overall, good job!

Comment (SindreSvendby) [reply src/transform/destructParam.js]: Good point, fixed!

File: src/transform/destructParam.js [hunk 2]
```diff
@@ -27,10 +27,7 @@ export default function(ast, logger) {
 
             // skip destructuring, if one of the destructured variable names is already declared by another destruction
             const duplicateDetected = exs.some(ex => {
-              if (variableNamesCreated.includes(ex.property.name)) {
-                return true;
-              }
-              return false;
+              return variableNamesCreated.includes(ex.property.name)
```
Comment (uniibu) [inline src/transform/destructParam.js]: travis failing because if missing semi-colon. 🎉

Comment (SindreSvendby) [reply src/transform/destructParam.js]: Thx! Fixed now! Have gotten too used to prettier!

**Comment by nene** [comment]
Hi @SindreSvendby, thanks for your efforts.

Unfortunately the fix for this isn't really quite as simple. Here's a test that will fail as a result of your changes:

```js
it('should transform the same param in multiple functions', () => {
  expectTransform(
    'function fn1(a) {\n' +
    '  console.log(a.foo);\n' +
    '}\n' +
    'function fn2(b) {\n' +
    '  console.log(b.foo);\n' +
    '}'
  ).toReturn(
    'function fn1({foo}) {\n' +
    '  console.log(foo);\n' +
    '}\n' +
    'function fn2({foo}) {\n' +
    '  console.log(foo);\n' +
    '}'
  );
});
```

This also partly explains why there aren't any issues labeled as good first issues - I've fixed all the simple things already. All the remaining issues are really not that simple.

**Comment by uniibu** [comment]
> This also partly explains why there aren't any issues labeled as good first issues - I've fixed all the simple things already. All the remaining issues are really not that simple.

That sounds a bit discouraging for someone who wants to contribute/help. Sorry, but It sounded like "all issues left here are complicated, nothing simple. If you don't do complicated leave" 

Anyhow, still a good try @SindreSvendby  Kudos!

**Comment by SindreSvendby** [comment]
Ok, spent two more minutes on this. 

So as far as I can see, a good implement here, would be a recursive `inScope` function that keep tracks on the identifiers declared in the scope. I naively thought  that the was what `withScope` was dooing. 
It seems like a general purpose helper that could be reused more places, but I'm way to new to this project to know if that in fact is the case.   

Please let me know if you think I'm going down the wrong rabbit hole here! 

Variables is not hoist and it's lexical scope, so should be able to identify the scope and it's identifiers, if not I guess I will learn something!

**Comment by nene** [comment]
Currently the fundamental problem is that while this `withScope()` helper gives us scope information about existing variables (so we can avoid conflicts when introducing new ones), but the information will not be updated and as soon as we introduce a new variable, things will get out of sync.

One way to overcome this problem would be to re-generate the scoping information each time a new variable is introduced. That would be tricky to do while in the middle of processing the arguments of one function - but I think for the single-function case we can implement a simple local cache of introduced variable names.

So, my suggestion would be to first solve the problem of handling the duplicates problem in this case:

```js
function fn(a, b) {
  console.log(a.foo, b.foo);
}
```

And then afterwards look into the harder problem of fixing the other test case.

**Comment by nene** [comment]
BTW. You're right about the class transform - it does need a rewrite to properly fix majority of its problems.

I looked through other issues, and found two possibly simpler issues to tackle:

- #186 super() calls semantics. While it is part of the class transform, this part can be worth implementing separately (we can likely carry that logic easily over when rewriting the class transform on one fine day).
- #266 new function edge case. Pretty odd corner case, but would be nice if we handled it.

One possibility is to look into issues labeled as "blocked externally" to see if they're still blocked.

**Comment by nene** [comment]
I'm closing down this old PR.

**Pull request closed without merging**